### PR TITLE
feat: Terraform example

### PR DESCRIPTION
**Pull request by skal111**
**Issue #1445:**

## Description of changes:

Added a core-example with Terraform deployment

Just added a new terraform folder under powertools-examples-core which contain appropriate instructions and template to deploy the sample Lambdas and API Gateway with Terraform

**Checklist**

* [ X ] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda-java/#tenets)
* [ N/A ] Update tests
* [ X ] Update docs
* [ ] PR title follows [conventional commit semantics](https://www.conventionalcommits.org/en/v1.0.0/)

## Breaking change checklist

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: examples/powertools-examples-core/terraform/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#  Powertools for AWS Lambda (Java) - Core Utilities Example with Terraform
+
+This project demonstrates the Lambda for Powertools Java module deployed using [Terraform](https://www.terraform.io/).
+For general information on the deployed example itself, you can refer to the parent [README](../README.md).
+To install Terraform if you don't have it yet, you can follow the [Install Terraform Guide](https://developer.hashicorp.com/terraform/downloads?product_intent=terraform).
+
+## Configuration
+Serverless Framework uses [serverless.yml](./serverless.yml) to define the application's AWS resources.
```
Comment (jeromevdl) [inline examples/powertools-examples-core/terraform/README.md:8]: terraform...

File: examples/powertools-examples-core/terraform/infra/api-gateway.tf [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+resource "aws_lambda_permission" "hello_world_lambda_invoke" {
+  statement_id  = "AllowAPIGatewayInvoke"
+  action        = "lambda:InvokeFunction"
+  function_name = "${aws_lambda_function.hello_world_lambda.function_name}"
+  principal     = "apigateway.amazonaws.com"
+  source_arn    = "${aws_api_gateway_rest_api.hello_world_api.execution_arn}/${aws_api_gateway_deployment.prod_deployment.stage_name}/GET/hello"
+}
+
+# Allows API gateway to invoke lambda
+resource "aws_lambda_permission" "hello_world_lambda_testinvoke" {
```
Comment (jeromevdl) [inline examples/powertools-examples-core/terraform/infra/api-gateway.tf:68]: what is this test-invoke-stage ?

File: examples/powertools-examples-core/terraform/infra/api-gateway.tf [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+resource "aws_lambda_permission" "hello_world_stream_lambda_invoke" {
+  statement_id  = "AllowAPIGatewayInvoke"
+  action        = "lambda:InvokeFunction"
+  function_name = "${aws_lambda_function.hello_world_stream_lambda.function_name}"
+  principal     = "apigateway.amazonaws.com"
+  source_arn    = "${aws_api_gateway_rest_api.hello_world_api.execution_arn}/${aws_api_gateway_deployment.prod_deployment.stage_name}/GET/hellostream"
+}
+
+# Allows API gateway to invoke lambda
+resource "aws_lambda_permission" "hello_world_stream_lambda_testinvoke" {
```
Comment (jeromevdl) [inline examples/powertools-examples-core/terraform/infra/api-gateway.tf:86]: same here?

File: examples/powertools-examples-core/terraform/infra/lambda.tf [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  handler           = "helloworld.App"
+  description       = "Powertools example, deployed by Terraform"
+  timeout           = 20
+  memory_size       = 512
+  role              = "${aws_iam_role.iam_role_for_lambda.arn}"
+  environment {
+    variables = {
+      POWERTOOLS_LOG_LEVEL = "INFO"
+      POWERTOOLS_LOGGER_SAMPLE_RATE = "0.1"
+      POWERTOOLS_LOGGER_LOG_EVENT = "true"
```
Comment (jeromevdl) [inline examples/powertools-examples-core/terraform/infra/lambda.tf:16]: remove this line, we actually don't handle this env var...

File: examples/powertools-examples-core/terraform/infra/lambda.tf [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  handler           = "helloworld.AppStream"
+  description       = "Powertools example, deployed by Terraform"
+  timeout           = 20
+  memory_size       = 512
+  role              = "${aws_iam_role.iam_role_for_lambda.arn}"
+  environment {
+    variables = {
+      POWERTOOLS_LOG_LEVEL = "INFO"
+      POWERTOOLS_LOGGER_SAMPLE_RATE = "0.7"
+      POWERTOOLS_LOGGER_LOG_EVENT = "true"
```
Comment (jeromevdl) [inline examples/powertools-examples-core/terraform/infra/lambda.tf:41]: remove this line too

File: examples/powertools-examples-core/terraform/infra/api-gateway.tf (hunk 2 above)
Comment (skal111) [reply examples/powertools-examples-core/terraform/infra/api-gateway.tf:68]: Required stage to test the lambda invocation from the API Gateway console. Can be removed, but CDK example deploys similar permissions, including that test-invoke-stage

File: examples/powertools-examples-core/terraform/infra/lambda.tf (hunk 4 above)
Comment (skal111) [reply examples/powertools-examples-core/terraform/infra/lambda.tf:16]: Terraform example is based on the CDK example configuration which include these environment variables. Are you sure?

Comment (jeromevdl) [reply examples/powertools-examples-core/terraform/infra/lambda.tf:16]: yes, we need to cleanup all the examples (#1444)

File: examples/powertools-examples-core/terraform/main.tf [hunk 6]
```diff
@@ -0,0 +1,4 @@
+# terraform modules
```
Comment (scottgerring) [inline examples/powertools-examples-core/terraform/main.tf:1]: How do we know that we haven't broken this - can we lint it as part of the build?

Comment (jeromevdl) [reply examples/powertools-examples-core/terraform/main.tf:1]: could we do a `terraform validate` at least ?

Comment (jeromevdl) [reply examples/powertools-examples-core/terraform/main.tf:1]: @scottgerring we should something similar for SAM and the others... `sam build` / `sam validate --lint --template template.yml`

Comment (skal111) [reply examples/powertools-examples-core/terraform/main.tf:1]: ok I will add some form of linting/validation

**Pull request closed without merging**